src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4979,5 +4794,797 @@ await foreach (var o in new Enumerable())
+                // (8,33): error CS8652: The feature 'extension GetAsyncEnumerator' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "new C()").WithArguments("extension GetAsyncEnumerator").WithLocation(8, 33)
+                );
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaInternalExtensions()
+        {
+            string source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+    }
+}
+public static class Extensions
+{
+    internal static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (8,33): warning CS0279: 'C' does not implement the 'async streams' pattern. 'Extensions.GetAsyncEnumerator(C)' is either static or not public.
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_PatternStaticOrInaccessible, "new C()").WithArguments("C", "async streams", "Extensions.GetAsyncEnumerator(C)").WithLocation(8, 33),
+                // (8,33): error CS8411: Asynchronous foreach statement cannot operate on variables of type 'C' because 'C' does not contain a suitable public instance definition for 'GetAsyncEnumerator'
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "new C()").WithArguments("C", "GetAsyncEnumerator").WithLocation(8, 33)
+                );
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaExtensionInInternalClass()
+        {
+            string source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+    }
+}
+internal static class Extensions
+{
+    public static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithMscorlib46(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "123");
+        }
```

Comment:
Please add some tests demonstrating:
* The behavior when `GetAsyncEnumerator` returns a type that does not fulfill the rest of the pattern correctly.
    * Also when the type returned has an explicit implementation of `IAsyncEnumerator` but has 
* The behavior when there's a `GetAsyncEnumerator` directly on the type that does not follow the correct pattern, and there's an extension `GetAsyncEnumerator` that does follow it correctly.
* The behavior of multiple `GetAsyncEnumerator` extensions.
   * When all of them have the incorrect number of parameters.
   * When there's no best match.
       * When multiple extensions take C.
       * When C implements I1 and I2, and there extension methods for both.
   * When one of them doesn't fulfill the pattern and one does.
   * When there are multiple overloads, one of which has a most specific receiver type.
      * Including when the most-specific receiver doesn't implement the pattern correctly.
   * When there are multiple overloads, one of which has the least number of optional parameters.
* With task-like types on MoveNextAsync
* With cancellationtoken passing (Julien blogged about that here: http://blog.monstuff.com/archives/2019/03/async-enumerables-with-cancellation.html).
* Test when the enumerator type is disposable, both via pattern and via IDisposable.

Generally, please look through the tests already in this file on pattern-based async enumerable and copy them for extension methods. I expect a number of these test suggestions will apply to the sync version as well, I won't relist them over there. I would expect to see IL asserted where appropriate.

Reply:
> Also when the type returned has an explicit implementation of IAsyncEnumerator but has

It looks like this sentence wasn't finished off?

Reply:
Fixed.

Reply:
> Also when the type returned has an explicit implementation of IAsyncEnumerator but also has an extension method.

Do you mean extension MoveNext? Because currently we don't check for extension foreach.

Reply:
> With cancellationtoken passing (Julien blogged about that here: http://blog.monstuff.com/archives/2019/03/async-enumerables-with-cancellation.html).

I don't believe it's possible to cancel a Pattern based GetAsyncEnumerator

Reply:
> Do you mean extension MoveNext? Because currently we don't check for extension MoveNext.

I mean something like this:

```cs
using System;
using System.Collections.Generic;
using System.Threading;

public class C : IAsyncEnumerable<int> {
    public void M() {
    }

    IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(CancellationToken cancellationToken)
    {
        throw new NotImplementedException();
    }
}

public static class CHelper
{
    public static IAsyncEnumerator<int> GetAsyncEnumerator(this C c) => throw null;
}
```

But it looks like you've added one, thanks.